src/t8_forest/t8_forest_search/t8_forest_search.hxx

```diff
@@ -0,0 +1,319 @@
+   * \param[in] tree_lindex_of_first_leaf The index of the first leaf in the tree.
+   */
+  void
+  search_recursion (const t8_locidx_t ltreeid, t8_element_t *element, const t8_eclass_scheme_c *ts,
+                    t8_element_array_t *leaf_elements, const t8_locidx_t tree_lindex_of_first_leaf);
+
+  /** \brief Checks if the search should stop due to empty queries.
+   * 
+   */
+  virtual bool
+  stop_due_to_queries ()
+    = 0;
+
+  /** \brief Checks an element during the search.
+   *
+   * This function is called for each element encountered during the search.
+   * It passes the arguments to the callback function provided by the user.
+   *
+   * \param[in] ltreeid The local tree ID of the current element.
+   * \param[in] element A pointer to the current element being processed.
+   * \param[in] is_leaf A bool indicating whether the current element is a leaf (non-zero) or not (zero).
+   * \param[in] leaf_elements A pointer to an array of leaf elements.
+   * \param[in] tree_leaf_index The index of the current leaf element within the tree.
+   *
+   * \return True if the search should continue, false otherwise.
+   */
+  virtual bool
+  check_element (const t8_locidx_t ltreeid, const t8_element_t *element, const bool is_leaf,
+                 const t8_element_array_t *leaf_elements, const t8_locidx_t tree_leaf_index)
+    = 0;
+
+  /** \brief Checks queries during the search.
+   * 
+   * This function is called to check queries during the search.
+   * It passes the arguments to the callback function provided by the user.
+   * 
+   * \param[in] new_active_queries A vector of indices of active queries.
+   * \param[in] ltreeid The local tree ID of the current element.
+   * \param[in] element A pointer to the current element being processed.
+   * \param[in] is_leaf A bool indicating whether the current element is a leaf (non-zero) or not (zero).
+   * \param[in] leaf_elements A pointer to an array of leaf elements.
+   * \param[in] tree_leaf_index The index of the current leaf element within the tree.
+   */
+  virtual void
+  check_queries (std::vector<size_t> &new_active_queries, const t8_locidx_t ltreeid, const t8_element_t *element,
+                 const bool is_leaf, const t8_element_array_t *leaf_elements, const t8_locidx_t tree_leaf_index)
+    = 0;
+};
+
+template <typename Udata = void>
+class t8_search: public t8_search_base {
+ public:
+  t8_search (t8_search_element_callback<Udata> element_callback, t8_forest_t forest = nullptr,
+             Udata *user_data = nullptr)
+    : t8_search_base (forest), user_data (user_data), element_callback (element_callback)
+  {
+  }
+
+  /** \brief Updates the user data associated with the object.
+   *
+   * This function sets the user data pointer to the provided Udata object.
+   *
+   * \param[in] udata A pointer to the Udata object to be set as the user data.
+   */
+  void
+  update_user_data (Udata *udata)
+  {
+    this->user_data = udata;
+  }
+
+  Udata *user_data;
+
```

Comment:
The whole part from here until the end of the file requires more comments.

Reply:
The documentation of the functions is given in the base-class, but I will add documentation to the implementation